src/hook_signature_check.h

```diff
@@ -0,0 +1,116 @@
+struct ___________________This_is_the_culprit_plugin_____________________
+  <Plugin__, true> {
+  static constexpr bool value = true;
+};
+
+// The above specialized HookSignaturesMatch plugin class is instanciated
+// based on two types of pointers to the plugin base class hook method and
+// the derived method. If both pointer types and thus the signatures equal,
+// the first or second specialization is instanciated,
+// leading to its compile time constant 'value' being defined as true.
+// Else the unspecialized version of the template class is instantiated,
+// causing 'value' to be defined as false.
+
+#define MATCH_HOOK_TYPE(HOOK_METHOD, PLUGIN)                                   \
+__NL__                                                                         \
+__NL__  {                                                                      \
+__NL__    /* Check it the two signatures equal. If not, the                    \
+__NL__    * user has implemented a method with the same name but a             \
+__NL__    * different (wrong) signature.                                       \
+__NL__    */                                                                   \
+__NL__    typedef HookSignaturesMatch<                                         \
+__NL__                       decltype(&KaleidoscopePlugin::HOOK_METHOD),       \
+__NL__                       decltype(&PLUGIN::HOOK_METHOD)                    \
+__NL__                  > Check;                                               \
+__NL__                                                                         \
+__NL__    static_assert(Check::value,                                          \
+__NL__     "*****************************************************************" \
+__NL__     "******************** READ THIS CAREFULLY ! **********************" \
+__NL__     "**************************************************************** " \
+__NL__     "One of your plugins implemented a hook \"" #HOOK_METHOD            \
+__NL__     "\" whose signature differs from the base class' hook signature. "  \
+__NL__     "Please check the compiler messages above or below for an hint "    \
+__NL__     "about which plugin is the culprit. "                               \
+__NL__     "Then check all the hook methods implemented by the plugin "        \
+__NL__     "and compare them to the respective hook methods of "               \
+__NL__     "base class KaleidscopePlugin. Note: Two function signatures "      \
+__NL__     "equal if all argument types and the method's "                     \
+__NL__     "const qualifier equal. "                                           \
+__NL__     "*****************************************************************" \
+__NL__     "*****************************************************************" \
+__NL__     "*****************************************************************" \
+__NL__    );                                                                   \
```

Comment:
That message should be hard to miss!

Reply:
I've tried it out, and it definitely makes the errors easy to find. It would be even better, however, with newlines in the error message string (move the closing quote after `#HOOK_METHOD` to the previous line, and "\n" to the beginning of each line, and another one at the end, and it becomes much easier to read).